### ISSUE TITLE
Python: KeyError in FindFileContainingSymbol

### ISSUE THREAD

**Issue by mehrdada** [google/protobuf#2882]
The attached code works fine on Linux (finds the File descriptor object) but results in KeyError on protobuf 3.2.0 on macOS [directly tested on Sierra 10.12.3, Python 2.7.10, but presumably Windows fails too, as our test infrastructure complains about not exactly this, but something that led me to this].

```
$ ./repro.sh
New python executable in /private/tmp/cleaner/demo/bin/python
Installing setuptools, pip, wheel...done.
Collecting protobuf==3.2.0
  Using cached protobuf-3.2.0-py2.py3-none-any.whl
Requirement already satisfied: six>=1.9 in ./demo/lib/python2.7/site-packages (from protobuf==3.2.0)
Requirement already satisfied: setuptools in ./demo/lib/python2.7/site-packages (from protobuf==3.2.0)
Requirement already satisfied: packaging>=16.8 in ./demo/lib/python2.7/site-packages (from setuptools->protobuf==3.2.0)
Requirement already satisfied: appdirs>=1.4.0 in ./demo/lib/python2.7/site-packages (from setuptools->protobuf==3.2.0)
Requirement already satisfied: pyparsing in ./demo/lib/python2.7/site-packages (from packaging>=16.8->setuptools->protobuf==3.2.0)
Installing collected packages: protobuf
Successfully installed protobuf-3.2.0
Traceback (most recent call last):
  File "bug.py", line 13, in <module>
    file_descriptor = pool.FindFileContainingSymbol(extension_descriptor.full_name)
  File "/private/tmp/cleaner/demo/lib/python2.7/site-packages/google/protobuf/descriptor_pool.py", line 304, in FindFileContainingSymbol
    raise error
KeyError: 'grpc.testing.proto2.Madeon'

$ ./demo/bin/python -m pip freeze
appdirs==1.4.3
packaging==16.8
protobuf==3.2.0
pyparsing==2.2.0
six==1.10.0
```

The following is the environment on which it WORKS (does NOT manifest the issue):

```
(demo) ~/repro$ python --version 
Python 2.7.12
(demo) ~/repro$ uname -a 
Linux mmx-saturn 4.4.0-67-generic #88-Ubuntu SMP Wed Mar 8 16:34:45 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
(demo) ~/repro$ python -m pip freeze
appdirs==1.4.3
enum34==1.1.6
futures==3.0.5
packaging==16.8
pkg-resources==0.0.0
protobuf==3.2.0
pyparsing==2.2.0
six==1.10.0
(demo) ~/repro$ virtualenv --version 
15.0.1
(demo) ~/repro$ python bug.py
<google.protobuf.pyext._message.FileDescriptor object at 0x7f2d88c84aa8>
```

(It is basically a GCE machine running Ubuntu 16.04 LTS and `sudo apt install unzip python-dev python-pip python-virtualenv`)

[KeyError-macOS-Windows.zip](https://github.com/google/protobuf/files/856374/KeyError-macOS-Windows.zip)

**Comment by nathanielmanistaatgoogle** [comment]
I'm seeing the problem on my Debian-in-GCE machine with
```
$ virtualenv --version
13.1.2
```
and
```
$ demo/bin/pip freeze
appdirs==1.4.3
packaging==16.8
protobuf==3.2.0
pyparsing==2.2.0
six==1.10.0
```
.

**Comment by mehrdada** [comment]
I have added a test via PR https://github.com/google/protobuf/pull/2887 (which currently fails when I run it) that tries to search for an extension via its full name. 

I also base this on the assumption that *Symbol* includes extension names, as nowhere in the documentation it is explicitly mentioned otherwise.

**Comment by xfxyjwf** [comment]
@anandolee to take a look

**Comment by anandolee** [comment]
Only top level extensions are included in Symbol. You test "python.internal.Factory2Message.one_more_field" is not a top level extension. Is grpc also this case?
Test can pass if you try "python.internal.another_field".

**Comment by mehrdada** [comment]
I see. It would be great if that restriction was [documented in the API](https://github.com/google/protobuf/blob/259dd7e842dd3a848a1094a42c05610e8a31dcbb/python/google/protobuf/descriptor_pool.py#L275). Is there a supported way to get the file descriptor representing any extension? We need such API to implement the [reflection service for gRPC](https://github.com/grpc/grpc/issues/2565). 

By the way, it is interesting that the exact same test would pass on a clean Ubuntu 16.04 LTS environment on GCE) with just  `sudo apt install unzip python-dev python-pip python-virtualenv` when you extract the zip file and run the `./repro.sh` file.

**Comment by anandolee** [comment]
Make sense to fix FindFileContainingSymbol be able to find none top extensions. Just want to confirm with you that if this can fix grpc's issue.

**Comment by mehrdada** [comment]
Right. Yes, we basically need a mechanism to map a type name and extension number to the file descriptor—as seen in this short function linked below, we are trying to use `FindFileContainingSymbol` now: https://github.com/grpc/grpc/pull/10216/commits/11473034da63f2c8309a4d40dde7554d2009b2ca

That is one solution to our problem. However, we are definitely open to suggestions if there is a better way to accomplish what we are trying to do.

**Comment by mehrdada** [comment]
I have dug further into the problem and realized why the behavior is platform dependent. It seems the root cause for the particular repro instructions I have attached here (which is similar to how we use it in gRPC reflection) is not necessarily that the extensions are not top-level [the particular extension attached to this post seems to be discoverable regardless], but it is correlated with that the python wheels on PyPI having the C extension compiled in them. 

On Mac, `pip` tries to fetch `protobuf-3.2.0-py2.py3-none-any.whl` and it doesn't have any `.so` files within itself. Likewise, on a default Goobuntu instance with an old pip version, the same file will be downloaded and it fails to work.  On an Ubuntu 16.04 LTS and when I explicitly upgrade `pip` in my virtual environment on Goobuntu, `pip` fetches `protobuf-3.2.0-cp27-cp27mu-manylinux1_x86_64.whl` which has `_message.so` along with python sources and it works just fine.

Is this a build/distribution issue?

**Comment by xfxyjwf** [comment]
See also b/36569071

**Comment by anandolee** [comment]
cl/152412285  cl/152543226

**Comment by xfxyjwf** [comment]
@mehrdada Jie's PR https://github.com/google/protobuf/pull/2962 should have fixed the problem. Note that the change is included only in the 3.3.x branch.
